src/dotnet/commands/dotnet-build/Program.cs

```diff
@@ -93,10 +107,55 @@ public static int Run(string[] args)
 
                 msbuildArgs.AddRange(app.RemainingArguments);
 
-                return new MSBuildForwardingApp(msbuildArgs).Execute();
+                ret._forwardingApp = new MSBuildForwardingApp(msbuildArgs);
+
+                return 0;
             });
 
-            return app.Execute(args);
+            int exitCode = app.Execute();
+            if (!codeExecuted)
+            {
+                throw new NonZeroExitCodeException(exitCode);
+            }
+
+            return ret;
+        }
+
+        public static int Run(string[] args)
+        {
+            DebugHelper.HandleDebugSwitch(ref args);
+
+            BuildCommand cmd;
+            try
+            {
+                cmd = FromArgs(args);
+            }
+            catch (NonZeroExitCodeException e)
+            {
+                return e.ExitCode;
+            }
+
+            return cmd.Execute();
+        }
+
+        public ProcessStartInfo GetProcessStartInfo()
+        {
+            return _forwardingApp.GetProcessStartInfo();
+        }
+
+        public int Execute()
+        {
+            return GetProcessStartInfo().Execute();
+        }
+
+        private class NonZeroExitCodeException : Exception
```

Comment:
likely misnamed - this is quite a hack around command line parser - i don't like it but no ideas how to implement this cleaner :(